src/linux/cgroups2.cpp

```diff
@@ -295,4 +296,152 @@ Try<set<string>> enabled(const string& cgroup)
 
 } // namespace controllers {
 
+namespace devices {
+
+const int ALLOW_ACCESS = 1;
+
+const int DENY_ACCESS = 0;
+
+// Utility class to construct an eBPF program to whitelist or blacklist
+// select device accesses.
+class DeviceProgram
+{
+public:
+  DeviceProgram() : program{ebpf::Program(BPF_PROG_TYPE_CGROUP_DEVICE)}
+  {
+    program.append({
+        // r2: Type ('c', 'b', '?')
+        BPF_LDX_MEM(BPF_W, BPF_REG_2, BPF_REG_1, 0),
+        BPF_ALU32_IMM(BPF_AND, BPF_REG_2, 0xFFFF),
+        // r3: Access ('r', 'w', 'm')
+        BPF_LDX_MEM(BPF_W, BPF_REG_3, BPF_REG_1, 0),
+        BPF_ALU32_IMM(BPF_RSH, BPF_REG_3, 16),
+        // r4: Major Version
+        BPF_LDX_MEM(BPF_W, BPF_REG_4, BPF_REG_1, 4),
+        // r5: Minor Version
+        BPF_LDX_MEM(BPF_W, BPF_REG_5, BPF_REG_1, 8)});
+  }
+
+  Try<Nothing> allow(const Entry entry) { return addDevice(entry, true); }
+
+  Try<Nothing> deny(const Entry entry) { return addDevice(entry, false); }
+
+  Try<Nothing> addDevice(const Entry entry, bool allow)
+  {
+    // We create a block of bytecode with the format:
+    // 1. [ Type Check          ] ⤵
+    // 2. [ Access Check        ] ⤵
```

Comment:
Hm.. why would we check access before we know that we're looking at the right device? i.e. that the device's type and major / minor match?

Reply:
Since all of the checks need to pass, I considered the order to be somewhat irrelevant. Ordering is semantically the same, we'd just short-circuit.

Reply:
I understand that regardless of ordering, these all need to be checked and we jump out if any don't match, but it seems more natural to reason about checking in the same order that we'd be writing the logic if we were to code the logic if we weren't generating assembly, in which case I would imagine that the code would be written to first examine if this is the right device, then check access. It also seems a bit easier to reason about the access part if it resides closer to the final access code?

Reply:
Reordered.